storage: prevent accidentally unpoisoning aborted txn

tbg opened the pull request:
Discovered by @bdarnell in
https://github.com/cockroachdb/cockroach/pull/18635#issuecomment-333393480.
Making poisoning happen less often (to reduce contention) is planned but
requires more care.

Closes #18635.

tbg committed:
storage: prevent accidentally unpoisoning aborted txn

Discovered by @bdarnell in https://github.com/cockroachdb/cockroach/pull/18635#issuecomment-333393480.
Making poisoning happen less often (to reduce contention) is planned but
requires more care.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/19093)
<!-- Reviewable:end -->

tbg committed:
storage: declare AbortCache only when used

This ensures that we won't declare the AbortSpan key for a txn
when committing or moving intents.

bdarnell commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 1 of 2 files at r1, 4 of 4 files at r2.
Review status: all files reviewed at latest revision, all discussions resolved, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/19093#-:-KwRf8osGUYoG3kk5vPD:bnfp4nl)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
TFTR! @bdarnell 1.1.1 or 1.1.2?

bdarnell commented:
Go ahead and cherry-pick it; looks like there's still time to put it in 1.1.1 (but I don't think it's a blocker for 1.1.1 since the issue has been around for a long time)
